### PR TITLE
chore: Update datafusion again

### PR DESCRIPTION
**Pull request by alamb**
# Rationale:
1. Keep up with dependencies
2. Get new fast grouping: https://github.com/apache/arrow-datafusion/pull/6904
3. Closes https://github.com/influxdata/influxdb_iox/issues/7282

# Changes

1. Update to DataFusion pin to https://github.com/apache/arrow-datafusion/commit/af26c66f708080e00967e41513c39c72e6bde8d1

### CODE REVIEW

File: iox_query_influxql/src/aggregate/percentile.rs [hunk 1]
```diff
@@ -39,7 +39,7 @@ pub(super) fn accumulator(dt: &DataType) -> Result<Box<dyn Accumulator>> {
 /// Calculate the intermediate merge state for the aggregator.
 pub(super) fn state_type(dt: &DataType) -> Result<Arc<Vec<DataType>>> {
     Ok(Arc::new(vec![
-        DataType::List(Arc::new(Field::new("state", dt.clone(), false))),
+        DataType::List(Arc::new(Field::new("item", dt.clone(), true))),
```
Comment (alamb) [inline iox_query_influxql/src/aggregate/percentile.rs:42]: the new grouping code verifies the types match, so updated this to match what is actually created `

Filed https://github.com/apache/arrow-rs/issues/4544 to track the idea of encapsulating this more

File: iox_query_influxql/src/plan/util_copy.rs [hunk 2]
```diff
@@ -145,34 +145,27 @@ where
                     negated,
                     expr,
                     pattern,
+                    case_insensitive,
```
Comment (alamb) [inline iox_query_influxql/src/plan/util_copy.rs:148]: due to https://github.com/apache/arrow-datafusion/pull/7007

File: iox_query_influxql/src/window/non_negative.rs [hunk 3]
```diff
@@ -21,17 +21,6 @@ struct NonNegative {
 }
 
 impl PartitionEvaluator for NonNegative {
-    fn update_state(
```
Comment (alamb) [inline iox_query_influxql/src/window/non_negative.rs:24]: Removed in https://github.com/apache/arrow-datafusion/pull/6966

**Comment by alamb** [comment]
@mhilton  @stuartcarnie  or @wolffcm  if one if you could double check the changes to the InfluxQL implementation that would be most appreciated. They are mostly due to API changes upstream (I think for the better)

**Review by mhilton — APPROVED** [review]
LGTM

**Comment by alamb** [comment]
Thanks @mhilton  and @wolffcm  

As I will be traveling tomorrow morning, I plan to merge this in probably Friday morning so I can be fully available for any blowback that happens

**Comment by alamb** [comment]
Given this PR has the potential to cause a rukus (as it has the new grouping code) I will be monitoring its rollout as well as prepare a rollback if necessary. 

See slack: https://influxdata.slack.com/archives/C04K1QRSDFG/p1689934331079049

**Comment by alamb** [comment]
Backup Revert PR: https://github.com/influxdata/influxdb_iox/pull/8304

**Pull request merged**